### PR TITLE
[DPB portsyncd/portmgrd/portorch] Support dynamic port add/deletion without dependencies

### PR DESCRIPTION
**Pull request by zhenggen-xu**
**What I did**
Support port add/delete cases in different daemons without dependencies

**Why I did it**
Baseline code for dynamic port breakout code.   

**How I verified it**
vs test cases

**Details if related**

Support port add/delete cases in different daemons
    
Changes were made on portmgrd/portsyncd and orchagent portsorch
so it should be able to remove/add ports in case no configuration
dependencies or runtime depencies (neighbor, mac etc) on them
    
Also skipped the netlink for port add/delete with master in portsyncd
and cleaned up g_init and g_portSet flag and data strcutures usage.
    
Added dynamic portbeakout test cases including the conf_test.py changs

Signed-off-by: Zhenggen Xu <zxu@linkedin.com>
Signed-off-by: Vasant Patil <vapatil@linkedin.com>

### CODE REVIEW

**Comment by zhenggen-xu** [comment]
retest this please

**Comment by zhenggen-xu** [comment]
retest this please!

File: orchagent/portsorch.cpp [hunk 1]
```diff
@@ -1550,6 +1555,25 @@ bool PortsOrch::initPort(const string &alias, const set<int> &lane_set)
     return true;
 }
 
+void PortsOrch::deInitPort(string alias, sai_object_id_t port_id)
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >deInitPort [](start = 16, length = 10)

removePortCounter ?

File: orchagent/portsorch.cpp [hunk 2]
```diff
@@ -1616,6 +1640,37 @@ void PortsOrch::cleanPortTable(const vector<string>& keys)
...
+
+    for (auto it = m_lanesAliasSpeedMap.begin(); it != m_lanesAliasSpeedMap.end();)
+    {
+        if (get<0>(it->second) == alias)
+        {
+            SWSS_LOG_NOTICE("Removing port %s from lanes map", alias.c_str());
+            it = m_lanesAliasSpeedMap.erase(it);
+            break;
+        }
+        it++;
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >it++ [](start = 8, length = 4)

move into `for` statement?

File: orchagent/portsorch.cpp [hunk 3]
```diff
@@ -1616,6 +1640,37 @@ void PortsOrch::cleanPortTable(const vector<string>& keys)
...
+
+    for (auto it = m_portListLaneMap.begin(); it != m_portListLaneMap.end();)
+    {
+        if (it->second == port_id)
+        {
+            SWSS_LOG_NOTICE("Removing port-id %lx from port list map", port_id);
+            it = m_portListLaneMap.erase(it);
+            break;
+        }
+        it++;
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >it++ [](start = 8, length = 4)

The same

File: orchagent/portsorch.cpp [hunk 4]
```diff
@@ -2108,9 +2163,30 @@ void PortsOrch::doPortTask(Consumer &consumer)
                 }
             }
         }
-        else
+        else if (op == DEL_COMMAND)
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >else if [](start = 8, length = 7)

After `else if`, you still need another `else`

File: orchagent/portsorch.cpp [hunk 5]
```diff
@@ -1432,6 +1433,7 @@ bool PortsOrch::addPort(const set<int> &lane_set, uint32_t speed, int an, string
     }
 
     m_portListLaneMap[lane_set] = port_id;
+    m_portCount++;
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >m_portCount [](start = 4, length = 11)

It is assigned in `PortsOrch::PortsOrch()` and unchanged later. Do you intend to change the semantics?

File: orchagent/portsorch.cpp [hunk 6]
```diff
@@ -1449,13 +1451,16 @@ bool PortsOrch::removePort(sai_object_id_t port_id)
...
 
+    removeAclTableGroup(p);
+
     sai_status_t status = sai_port_api->remove_port(port_id);
     if (status != SAI_STATUS_SUCCESS)
     {
         SWSS_LOG_ERROR("Failed to remove port %" PRIx64 ", rv:%d", port_id, status);
         return false;
     }
-    removeAclTableGroup(p);
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >removeAclTableGroup [](start = 4, length = 19)

Why move it? Was there a bug?

File: orchagent/portsorch.cpp [hunk 7]
```diff
@@ -1776,7 +1831,7 @@ void PortsOrch::doPortTask(Consumer &consumer)
              * 2. Create new ports
              * 3. Initialize all ports
              */
-            if (m_portConfigState == PORT_CONFIG_RECEIVED && (m_lanesAliasSpeedMap.size() == m_portCount))
+            if (m_portConfigState == PORT_CONFIG_RECEIVED || m_portConfigState == PORT_CONFIG_DONE)
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >if (m_portConfigState == PORT_CONFIG_RECEIVED || m_portConfigState == PORT_CONFIG_DONE) [](start = 12, length = 87)

This change is tricky. Please get approval from

File: portsyncd/linksync.cpp [hunk 8]
```diff
@@ -209,6 +211,12 @@ void LinkSync::onMsg(int nlmsg_type, struct nl_object *obj)
         return;
     }
 
+    /* If this is a bridge port, return */
```
Comment (qiluo-msft) [inline portsyncd/linksync.cpp]: >bridge port [](start = 20, length = 11)

The comment says differently than the code.

File: portsyncd/linksync.cpp [hunk 9]
```diff
@@ -226,27 +234,24 @@ void LinkSync::onMsg(int nlmsg_type, struct nl_object *obj)
     /* Insert or update the ifindex to key map */
     m_ifindexNameMap[ifindex] = key;
 
+    if (nlmsg_type == RTM_DELLINK)
+    {
+        m_statePortTable.del(key);
+        SWSS_LOG_NOTICE("Delete %s(ok) from state db", key.c_str());
```
Comment (qiluo-msft) [inline portsyncd/linksync.cpp]: >(ok) [](start = 34, length = 4)

Why you add `ok` ?

File: portsyncd/linksync.cpp [hunk 10]
```diff
@@ -226,27 +234,24 @@ void LinkSync::onMsg(int nlmsg_type, struct nl_object *obj)
     /* Insert or update the ifindex to key map */
     m_ifindexNameMap[ifindex] = key;
 
+    if (nlmsg_type == RTM_DELLINK)
+    {
+        m_statePortTable.del(key);
```
Comment (qiluo-msft) [inline portsyncd/linksync.cpp]: >m_statePortTable [](start = 8, length = 16)

Will you also remove it from m_portTable?

File: portsyncd/linksync.cpp [hunk 11]
```diff
@@ -226,27 +234,24 @@ void LinkSync::onMsg(int nlmsg_type, struct nl_object *obj)
...
     if (m_portTable.get(key, temp))
     {
-        /* TODO: When port is removed from the kernel */
-        if (nlmsg_type == RTM_DELLINK)
-        {
-            return;
-        }
-
-        /* Host interface is created */
-        if (!g_init && g_portSet.find(key) != g_portSet.end())
```
Comment (qiluo-msft) [inline portsyncd/linksync.cpp]: >if (!g_init && g_portSet.find(key) != g_portSet.end()) [](start = 8, length = 54)

I think this check is for restart swss case. Did you tested this case?

File: tests/conftest.py [hunk 12]
```diff
@@ -77,8 +77,8 @@ def __init__(self, dvs):
 
 class ApplDbValidator(object):
     def __init__(self, dvs):
-        appl_db = swsscommon.DBConnector(swsscommon.APPL_DB, dvs.redis_sock, 0)
-        self.neighTbl = swsscommon.Table(appl_db, "NEIGH_TABLE")
+        self.appl_db = swsscommon.DBConnector(swsscommon.APPL_DB, dvs.redis_sock, 0)
+        self.neighTbl = swsscommon.Table(self.appl_db, "NEIGH_TABLE")
```
Comment (qiluo-msft) [inline tests/conftest.py]: Many changes are not related to DPB at all. Could you help create another PR? We may merge it quickly and cleanly.

**Comment by lguohan** [comment]
retest this please

**Comment by lguohan** [comment]
LGTM

**Comment by lguohan** [comment]
can you check why build fails?

**Comment by zhenggen-xu** [comment]
> can you check why build fails?

Need rework the code due to PR 1170 ( https://github.com/Azure/sonic-swss/pull/1170 ). New code pushed.

File: portsyncd/linksync.cpp (hunk 11 above)
Comment (qiluo-msft) [reply portsyncd/linksync.cpp]: The steps are:
1. everything is running
2. restart swss
3. during create_switch, SAI will remove all the host interfaces and recreate them
4. the remove messages will reach here, and you need to ignore these messages

@prsunny may clarify if anything wrong.

---
In reply to: [359622619](https://github.com/Azure/sonic-swss/pull/1112#discussion_r359622619) [](ancestors = 359622619)

Comment (zhenggen-xu) [reply portsyncd/linksync.cpp]: @qiluo-msft @prsunny  The remove messages was done above, will not reach here
```
    if (nlmsg_type == RTM_DELLINK)
    {
        m_statePortTable.del(key);
        SWSS_LOG_NOTICE("Delete %s(ok) from state db", key.c_str());
        return;
    }
```

I think for that case, we should be fine.

**Comment by zhenggen-xu** [comment]
retest this please

**Comment by zhenggen-xu** [comment]
retest this please

**Comment by daall** [comment]
retest this please

**Comment by daall** [comment]
retest this please

**Comment by daall** [comment]
retest this please

**Comment by daall** [comment]
retest this please

**Comment by daall** [comment]
retest this please

**Comment by zhenggen-xu** [comment]
retest this please

**Comment by lguohan** [comment]
retest this please

**Comment by zhenggen-xu** [comment]
retest this please

File: portsyncd/linksync.cpp (hunk 11 above)
Comment (prsunny) [reply portsyncd/linksync.cpp]: @zhenggen-xu , as Qi mentioned, this is for the restart swss case. There would be a period where the previous interfaces are present in kernel before SAI creates the new interfaces and you may get some netlink notifications during swss start. Since the check is removed, you would be adding it to state_db immediately. This is my thought!

Comment (zhenggen-xu) [reply portsyncd/linksync.cpp]: Based on the service file, if we restart swss service,  we stop swss docker, stop syncd docker.  Then start swss docker, start syncd docker.  It is true that swss daemons are running when syncd was starting, however,  syncd should be deleting the interfaces then add them, right?  How is it going to produce RTM_NEWLINK if it is not really creating the interface which will trigger this code to add to state_db?

Comment (zhenggen-xu) [reply portsyncd/linksync.cpp]: Also, we do have the code above to filter the old interfaces messages if any:
```
    if (m_ifindexOldNameMap.find(ifindex) != m_ifindexOldNameMap.end())
    {
        SWSS_LOG_INFO("nlmsg type:%d Ignoring message for old interface %s(%d)",
                nlmsg_type, key.c_str(), ifindex);
        return;
    }
```
